Preventing actions called from another action to show in summary

marcelofabri opened the pull request:
Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes:
- [x] Run `rspec` for all tools you modified
- [x] Run `rubocop -a` to ensure the code style is valid
- [x] We currently don't accept new actions, please publish a plugin instead, more information in [Plugins.md](https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Plugins.md)

Before submitting a pull request, we appreciate if you create an issue first to discuss the change :+1:

This should fix #4546. We probably should add at least a test to make sure we don't have regressions on this, but just wanted to see it the implementation makes sense first.

Just to make clear, this hides the "inner actions" from the summary:

```
+------+-------------------------------------+-------------+
|                     fastlane summary                     |
+------+-------------------------------------+-------------+
| Step | Action                              | Time (in s) |
+------+-------------------------------------+-------------+
| 1    | opt_out_usage                       | 0           |
| 2    | Verifying required fastlane version | 0           |
| 3    | default_platform                    | 0           |
| 4    | is_ci                               | 0           |
| 5    | bundle_install                      | 0           |
| 6    | is_ci                               | 0           |
| 7    | cocoapods_if_needed                 | 98          |
+------+-------------------------------------+-------------+
```

You can compare to the behavior before this PR below:

```
$ fastlane pod_install

...
+------+-------------------------------------+-------------+
|                     fastlane summary                     |
+------+-------------------------------------+-------------+
| Step | Action                              | Time (in s) |
+------+-------------------------------------+-------------+
| 1    | opt_out_usage                       | 0           |
| 2    | Verifying required fastlane version | 0           |
| 3    | default_platform                    | 0           |
| 4    | is_ci                               | 0           |
| 5    | bundle_install                      | 0           |
| 6    | is_ci                               | 0           |
| 7    | cocoapods                           | 51          |
| 8    | cocoapods_if_needed                 | 62          |
+------+-------------------------------------+-------------+
```

marcelofabri committed:
Preventing actions called from another actino to show in summary

Check #4546 for more details

KrauseFx commented:
Hey @marcelofabri, sorry for the late reply on this, and thanks for your pull request. I don't think we can merge the pull request as it is now. We'd need some more documentation in the code, and also tests covering this change. 
Is that something you could do?

marcelofabri commented:
No worries! Where exactly do you want to see more docs? In `trigger_action_by_name` header?

Also, can you point me to where this is being tested or how could I add tests for this? (Since it needs to capture the summary output)

ohwutup commented:
Hey @marcelofabri - Apologies for the delay here- I think a good place to write a test for this would be in `fastlane_runner_spec.rb`. And Regarding adding more docs to the code, I think if you put a little comment above `trigger_action_by_name`'s definition about what the `from_action` param is doing, it might be helpful. Thanks! 🐙

marcelofabri commented:
@ohwutup I've rebased this and added a comment in the header, but I'm still having trouble on how to test this without mocking the whole stack. Are there any integration tests? Maybe this would  be better tested in that way.

marcelofabri committed:
Add tests 🚀

marcelofabri commented:
@ohwutup So, I took another look and found `action_spec`. I was able to add tests there in e4ff4d6a. I'm not sure if it's the best approach, but the new tests do fail without 52bfe29 and pass with it. 🎉

milch commented on fastlane/lib/fastlane/runner.rb:
```diff
@@ -109,7 +109,9 @@ def find_alias(action_name)
-    def trigger_action_by_name(method_sym, custom_dir, *arguments)
+    # @param from_action Indicates if this action is being trigged by another action.
+    #                    If so, it won't show up in summary.
+    def trigger_action_by_name(method_sym, custom_dir, from_action, *arguments)
```
In the past we've had trouble with adding parameters to existing functions breaking people's setups, so could you give this a default value please? e.g. `def trigger_action_by_name(method_sym, custom_dir, from_action: false, *arguments)`

marcelofabri replied:
Sure, although I'd say that this method shouldn't really be called by end users 😅

marcelofabri replied:
Actually, it seems that this is not valid syntax because of the `*arguments` later 😬

milch replied:
> Sure, although I'd say that this method shouldn't really be called by end users 😅

That's never stopped someone from doing things before 😉 

> Actually, it seems that this is not valid syntax because of the *arguments later 😬

Damn, you're right. I just double checked that you changed all occurences of trigger_action_by_name, so we should be good.
